packages/ens-utils/src/registrar.ts

```diff
@@ -0,0 +1,217 @@
+  price: Price;
+  reason: string;
+  validity: TimePeriod;
+}
+
+export interface RegistrarServiceFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.ServiceFee;
+  price: Price;
+  reason?: string;
+  validity?: undefined;
+}
+
+export type RegistrarCharge = RegistrarBaseFee | RegistrarSpecialNameFee | RegistrarTemporaryFee | RegistrarServiceFee;
+
+/**
+ * A price quote from a `Registrar` to register or renew a domain for a given
+ * period of time.
+ */
+export interface AbstractRegistrarPriceQuote {
+
+  /**
+   * The action associated with this `RegistrarPriceQuote`.
+   */
+  action: RegistrarAction;
+
+  /**
+   * The `TimePeriod` that this `RegistrarPriceQuote` is for.
+   */
+  rentalPeriod: TimePeriod;
+
+  /**
+   * The set of distinct `RegistrarCharge` values that must be paid to the
+   * `Registrar` to rent the domain for the period of `rentalPeriod`.
+   * 
+   * May be empty if the domain is free to rent for `rentalPeriod`.
+   */
+  charges: RegistrarCharge[];
+
+  /**
+   * The total price to rent the domain for the period of `rentalPeriod`.
+   * 
+   * This is the sum of `charges` in the unit of currency that should be paid
+   * to the registrar.
+   */
+  totalPrice: Price;
+}
+
+export interface RegistrationPriceQuote extends AbstractRegistrarPriceQuote {
+  action: typeof RegistrarAction.Register;
+}
+
+export interface RenewalPriceQuote extends AbstractRegistrarPriceQuote {
+  action: typeof RegistrarAction.Renew;
+}
+
+export interface Registrar {
+  getManagedSubname(name: ENSName): ENSName | null;
+
+  // NOTE: Throws RegistrarUnsupportedNameError if name is not supported by this registrar
+  getValidatedSubname(name: ENSName): ENSName;
+
+  getOnchainRegistrar(): ContractRef | null;
+
+  canRegister(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration?: Registration): boolean;
+  getRegistrationPriceQuote(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration?: Registration): RegistrationPriceQuote;
+  
+  canRenew(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration: Registration): boolean;
+  getRenewalPriceQuote(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration: Registration): RenewalPriceQuote;
+}
+
+export interface OnchainRegistrar extends Registrar {
+  getOnchainRegistrar(): ContractRef;
+}
+
+export interface OffchainRegistrar extends Registrar {
+  getOnchainRegistrar(): null;
+}
+
+/**
+ * Identifies that a name was passed to a `Registrar` function that was not relevant for that `Registrar`.
+ * 
+ * @param message a reason why `name` was NOT relevant for `Registrar`.
+ * @param label the `ENSName` value that was NOT relevant.
+ */
+export class RegistrarUnsupportedNameError extends Error {
+
+  public constructor(message: string, name: ENSName) {
+    super(`RegistrarUnsupportedNameError for name "${name.name}": ${message}`);
+    this.name = "RegistrarUnsupportedNameError";
```

Comment:
Might be good to exchange `name` convention so we don't have a errorName as string and a ensName as ENSName?

Reply:
I don't understand?

Reply:
Both `constructor` `name` param *and* `RegistrarUnsupportedNameError` class `name` prop are defined as `name`... in the first case, a class constructor param is defined as `ENSName`, while, in the second one, the class public variable is defined as `string`.

My suggestion is to rename the class constructor `name` to `ensName` to avoid these duplicates